codex/erasure/asyncbackend.nim

```diff
@@ -0,0 +1,191 @@
+## those terms.
+
+import std/sequtils
+
+import pkg/taskpools
+import pkg/chronos
+import pkg/chronos/threadsync
+import pkg/questionable/results
+
+import ./backend
+import ../errors
+import ../logutils
+
+logScope:
+  topics = "codex asyncerasure"
+
+type
+  EncoderBackendPtr = ptr EncoderBackend
+  DecoderBackendPtr = ptr DecoderBackend
+
+  # Args objects are missing seq[seq[byte]] field, to avoid unnecessary data copy
+  EncodeTaskArgs = object
+    signal: ThreadSignalPtr
+    backend: EncoderBackendPtr
+    blockSize: int
+    ecM: int
+
+  DecodeTaskArgs = object
+    signal: ThreadSignalPtr
+    backend: DecoderBackendPtr
+    blockSize: int
+    ecK: int
+
+  SharedArrayHolder*[T] = object
+    data: ptr UncheckedArray[T]
+    size: int
+
+  EncodeTaskResult = Result[SharedArrayHolder[byte], cstring]
+  DecodeTaskResult = Result[SharedArrayHolder[byte], cstring]
+
+proc encodeTask(args: EncodeTaskArgs, data: seq[seq[byte]]): EncodeTaskResult =
+  var
+    data = data.unsafeAddr
+    parity = newSeqWith[seq[byte]](args.ecM, newSeq[byte](args.blockSize))
+
+  try:
+    let res = args.backend[].encode(data[], parity)
+
+    if res.isOk:
+      let
+        resDataSize = parity.len * args.blockSize
+        resData = cast[ptr UncheckedArray[byte]](allocShared0(resDataSize))
+        arrHolder = SharedArrayHolder[byte](
+          data: resData,
+          size: resDataSize
+        )
+
+      for i in 0..<parity.len:
+        copyMem(addr resData[i * args.blockSize], addr parity[i][0], args.blockSize)
+
+      return ok(arrHolder)
+    else:
+      return err(res.error)
+  except Exception as exception:
+    return err(exception.msg.cstring)
+  finally:
+    if err =? args.signal.fireSync().mapFailure.errorOption():
+      error "Error firing signal", msg = err.msg
+
+proc decodeTask(args: DecodeTaskArgs, data: seq[seq[byte]], parity: seq[seq[byte]]): DecodeTaskResult =
+  var
+    data = data.unsafeAddr
+    parity = parity.unsafeAddr
+    recovered = newSeqWith[seq[byte]](args.ecK, newSeq[byte](args.blockSize))
+
+  try:
+    let res = args.backend[].decode(data[], parity[], recovered)
+
+    if res.isOk:
+      let
+        resDataSize = recovered.len * args.blockSize
+        resData = cast[ptr UncheckedArray[byte]](allocShared0(resDataSize))
+        arrHolder = SharedArrayHolder[byte](
+          data: resData,
+          size: resDataSize
+        )
+
+      for i in 0..<recovered.len:
+        copyMem(addr resData[i * args.blockSize], addr recovered[i][0], args.blockSize)
+
+      return ok(arrHolder)
+    else:
+      return err(res.error)
+  except Exception as exception:
+    return err(exception.msg.cstring)
+  finally:
+    if err =? args.signal.fireSync().mapFailure.errorOption():
+      error "Error firing signal", msg = err.msg
+
+proc proxySpawnEncodeTask(
+  tp: Taskpool,
+  args: EncodeTaskArgs,
+  data: ref seq[seq[byte]]
+): Flowvar[EncodeTaskResult] =
+  tp.spawn encodeTask(args, data[])
```

Comment:
> So two things to keep in mind, a) we aren't really making a copy of the data just the seq descriptor and b) this is something to keep in mind when debugging potential concurrency issues.

Generally that's the case. Taskpools wrap arguments in `Isolate` which creates a deepcopy unless it can do a move. However, that may not alleviate the concurrency issues with `refc` since it's not clear if the *ownership* of the copied seq data is transferred to the other thread. @tbekas and I tested that it is doing a deep copy, but didn't determine who ends up owning the seq.

```nim
  let taskNode = new(TaskNode, workerContext.currentTask) do:
    type
      ScratchObj_838861196 = object
        data: seq[seq[char]]
        sig: ThreadSignalPtr

    let scratch_838861188 = cast[ptr ScratchObj_838861196](c_calloc(csize_t(1),
        csize_t(16)))
    if isNil(scratch_838861188):
      raise
        (ref OutOfMemDefect)(msg: "Could not allocate memory", parent: nil)
    block:
      var isoTemp_838861192 = isolate(data)
      scratch_838861188.data = extract(isoTemp_838861192)
      var isoTemp_838861194 = isolate(sig)
      scratch_838861188.sig = extract(isoTemp_838861194)
    proc worker_838861197(args`gensym27: pointer) {.gcsafe, nimcall, raises: [].} =
      let objTemp_838861191 = cast[ptr ScratchObj_838861196](args`gensym27)
      let data_838861193 = objTemp_838861191.data
      let sig_838861195 = objTemp_838861191.sig
      worker(data_838861193, sig_838861195)

    proc destroyScratch_838861198(args`gensym27: pointer) {.gcsafe, nimcall,
        raises: [].} =
      let obj_838861199 = cast[ptr ScratchObj_838861196](args`gensym27)
      `=destroy`(obj_838861199[])

    Task(callback: worker_838861197, args: scratch_838861188,
         destroy: destroyScratch_838861198)
  schedule(workerContext, taskNode)
```

Reply:
All of that to say that using `ptr seq[seq[byte]]` like @dryajov suggests shouldn't cause us any more issues when using `refc` than using `seq[seq[byte]]` since it's unclear how `refc` mixes with `move`. In the best case @tbekas setup using `seq[seq[byte]]` would move the data to the foreign thread like it would in `arc`.

Reply:
Though I'm still concerned of why we're not getting foreign GC heap contamination issue like we were previously @dryajov. Seems like seq's don't cause that corruption to occur anymore. Perhaps the compiler uses `allocShared` now instead? But I've run a bunch of tests with seqs and don't seem to run into the GC-heap contamination issue anymore.

Reply:
@dryajov 
> we aren't really making a copy of the data just the seq descriptor

If that would be the case, the following test would fail, but it doesn't.

```nim
var tp = Taskpool.new(num_threads = 2)

proc runTask(zeroes: seq[byte]) =
  var ones = @[byte 1, 1, 1, 1]
  copyMem(unsafeAddr zeroes[0], addr ones[0], 4)

test "some test":
  var zeroes = newSeq[byte](4)

  tp.spawn runTask(zeroes)

  tp.syncAll()

  check:
    zeroes == @[byte 0, 0, 0, 0]
```

Reply:
It would be helpful to print out the expanded version of this, since this is all transformed with macros...

But specifically, with your example, I don't think it would fail either way because you're performing `raw` operations on a memory segment that is big enough to hold the values you're copying into it. 

So, to separate this things a bit, your example above, uses pointer operations, which the GC isn't aware off and since the allocated memory segment is big enough to hold the elements being copied onto it, it will just work.

If you would do a copy directly into `zeroes = @[1, 1, 1, 1]`, it would also work, provided that `zeroes` is big enough to hold all the `ones` being copied. If you would instead copy 5 ones, and zeroes can only hold 4, the GC __could__ move your seq elsewhere, because it would need to do a realloc or some combination of memmove/copy.

What the isolate adds to this picture, is that it simply moves the data onto the taskpools worker object, but the object still lives on the owning thread, not on the worker thread.

So, bottom line, the snippet above doesn't really demonstrate anything with regards to my description of the seq header. 

What I don't understand is why your claiming that it would fail at all and how is this relevant?

Reply:
@elcritch 

> All of that to say that using ptr seq[seq[byte]] like @dryajov suggests shouldn't cause us any more issues when using refc than using seq[seq[byte]] since it's unclear how refc mixes with move. In the best case @tbekas setup using seq[seq[byte]] would move the data to the foreign thread like it would in arc.

I don't think the move happens across GCs/threads, this is a move on to the holder object, that still lives on the owning thread, not the worker thread?

Reply:
I'm curious to see where the deep copy happens?

Reply:
>What I don't understand is why your claiming that it would fail at all and how is this relevant?

Because you said that:

>we aren't really making a copy of the data just the seq descriptor

which implies that we refer the same memory region from both threads. My example shows that threads refer their own copies. If you meant something else by this statement, please explain what you meant exactly.

Reply:
> > What I don't understand is why your claiming that it would fail at all and how is this relevant?
> 
> Because you said that:
> 
> > we aren't really making a copy of the data just the seq descriptor
> 
> which implies that we refer the same memory region from both threads. My example shows that threads refer their own copies, both descriptor and the data. If you meant something else by this statement, please explain what you meant exactly.

No, your example doesn't show that at all, that is the point I'm making in https://github.com/codex-storage/nim-codex/pull/716#discussion_r1530745733. For reference, you can see the transformed version of the encode/decode tasks.

I annotated the encode task with comments at the relevant places, decode follows the same patter, but it's here for completenes

Encode:

```nim
  let fut = newFlowVar(typeof(EncodeTaskResult))
  proc taskpool_encodeTask(args: EncodeTaskArgs; data: seq[seq[byte]];
                           fut: Flowvar[EncodeTaskResult]) {.nimcall.} =
    let res`gensym105 = encodeTask(args, data)
    readyWith(fut, res`gensym105)

  let task =
    type
      ScratchObj_822084605 = object
        args: EncodeTaskArgs
        data: seq[seq[byte]]
        fut: Flowvar[EncodeTaskResult]

    let scratch_822084595 = cast[ptr ScratchObj_822084605](c_calloc(csize_t(1),
        csize_t(48)))
    if isNil(scratch_822084595):
      raise
        (ref OutOfMemDefect)(msg: "Could not allocate memory", parent: nil)
    block:
      var isoTemp_822084599 = isolate(args)
      scratch_822084595.args = extract(isoTemp_822084599)
      var isoTemp_822084601 = isolate(data[])
      scratch_822084595.data = extract(isoTemp_822084601) # <---- the seq gets moved over, `extract` does `move` on the contained isolate
      var isoTemp_822084603 = isolate(fut)
      scratch_822084595.fut = extract(isoTemp_822084603)
    proc taskpool_encodeTask_822084606(args`gensym129: pointer) {.gcsafe,
        nimcall.} =
      let objTemp_822084598 = cast[ptr ScratchObj_822084605](args`gensym129)
      let args_822084600 = objTemp_822084598.args
      let data_822084602 = objTemp_822084598.data
      let fut_822084604 = objTemp_822084598.fut
      taskpool_encodeTask(args_822084600, data_822084602, fut_822084604)

    proc destroyScratch_822084607(args`gensym129: pointer) {.gcsafe, nimcall.} =
      let obj_822084608 = cast[ptr ScratchObj_822084605](args`gensym129)
      `=destroy`(obj_822084608[])

    Task(callback: taskpool_encodeTask_822084606, args: scratch_822084595,
         destroy: destroyScratch_822084607)
  let taskNode = new(TaskNode, workerContext.currentTask, task)


  schedule(workerContext, taskNode) # this is where everything gets scheduled on the worker thread, everything else happens in the owning thread
```

Decode:

```nim
  let fut = newFlowVar(typeof(DecodeTaskResult))
  proc taskpool_decodeTask(args: DecodeTaskArgs; data: seq[seq[byte]];
                           parity: seq[seq[byte]];
                           fut: Flowvar[DecodeTaskResult]) {.nimcall.} =
    let res`gensym133 = decodeTask(args, data, parity)
    readyWith(fut, res`gensym133)

  let task =
    type
      ScratchObj_822084876 = object
        args: DecodeTaskArgs
        data: seq[seq[byte]]
        parity: seq[seq[byte]]
        fut: Flowvar[EncodeTaskResult]

    let scratch_822084862 = cast[ptr ScratchObj_822084876](c_calloc(csize_t(1),
        csize_t(56)))
    if isNil(scratch_822084862):
      raise
        (ref OutOfMemDefect)(msg: "Could not allocate memory", parent: nil)
    block:
      var isoTemp_822084868 = isolate(args)
      scratch_822084862.args = extract(isoTemp_822084868)
      var isoTemp_822084870 = isolate(data[])
      scratch_822084862.data = extract(isoTemp_822084870)
      var isoTemp_822084872 = isolate(parity[])
      scratch_822084862.parity = extract(isoTemp_822084872)
      var isoTemp_822084874 = isolate(fut)
      scratch_822084862.fut = extract(isoTemp_822084874)
    proc taskpool_decodeTask_822084877(args`gensym138: pointer) {.gcsafe,
        nimcall.} =
      let objTemp_822084867 = cast[ptr ScratchObj_822084876](args`gensym138)
      let args_822084869 = objTemp_822084867.args
      let data_822084871 = objTemp_822084867.data
      let parity_822084873 = objTemp_822084867.parity
      let fut_822084875 = objTemp_822084867.fut
      taskpool_decodeTask(args_822084869, data_822084871, parity_822084873, fut_822084875)

    proc destroyScratch_822084878(args`gensym138: pointer) {.gcsafe, nimcall.} =
      let obj_822084879 = cast[ptr ScratchObj_822084876](args`gensym138)
      `=destroy`(obj_822084879[])

    Task(callback: taskpool_decodeTask_822084877, args: scratch_822084862,
         destroy: destroyScratch_822084878)
  let taskNode = new(TaskNode, workerContext.currentTask, task)
  schedule(workerContext, taskNode)
```

Reply:
In your code sample:

```nim
var isoTemp_822084601 = isolate(data[])
scratch_822084595.data = extract(isoTemp_822084601) # <---- the seq gets moved over, `extract` does `move` on the contained isolate
```

The second line maybe does the move, but the first one does the copy.

Reply:
IIRC, `var isoTemp_822084601 = isolate(data[])` performs a deepcopy. I've used `Isolate` before and currently it's conservative and clones the data.

Reply:
> @elcritch
> 
> > All of that to say that using ptr seq[seq[byte]] like @dryajov suggests shouldn't cause us any more issues when using refc than using seq[seq[byte]] since it's unclear how refc mixes with move. In the best case @tbekas setup using seq[seq[byte]] would move the data to the foreign thread like it would in arc.
> 
> I don't think the move happens across GCs/threads, this is a move on to the holder object, that still lives on the owning thread, not the worker thread?

@dryajov for `refc` I believe you're correct, assuming move in `refc` *doesn't* move the data ownership of the seq to the new thread. I think that makes sense, but then in theory we should be getting "heap contamination".

I was digging into the seq memory handling in `refc` with moves but didn't get too far.